agent/cmd/serve.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright © 2024 Pulumi Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"syscall"
+
+	grpc_zap "github.com/grpc-ecosystem/go-grpc-middleware/logging/zap"
+	grpc_ctxtags "github.com/grpc-ecosystem/go-grpc-middleware/tags"
+	pb "github.com/pulumi/pulumi-kubernetes-operator/agent/pkg/proto"
+	"github.com/pulumi/pulumi-kubernetes-operator/agent/pkg/server"
+	"github.com/pulumi/pulumi/sdk/v3/go/auto"
+	"github.com/spf13/cobra"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapio"
+	"google.golang.org/grpc"
+)
+
+var (
+	WorkDir     string
+	SkipInstall bool
+	Stack       string
+	Host        string
+	Port        int
+)
+
+var rpcLogger *zap.Logger
+
+// serveCmd represents the serve command
+var serveCmd = &cobra.Command{
+	Use:   "serve",
+	Short: "Serve the agent RPC service",
+	Long: `Start the agent gRPC server.
+`,
+	PreRun: func(cmd *cobra.Command, args []string) {
+		rpcLogger = zap.L().Named("rpc")
+		grpc_zap.ReplaceGrpcLoggerV2(rpcLogger)
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		ctx := cmd.Context()
```

Comment:
Not something you need to do now, but suggest eventually making these cobra functions pass-throughs to your server package. There's some non-trivial setup that's worth testing eventually, and it'll be easier to do that without the cobra overhead.